pkg/chains/storage/grafeas/grafeas.go

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2020 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grafeas
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	attestationpb "github.com/grafeas/grafeas/proto/v1beta1/attestation_go_proto"
+	commonpb "github.com/grafeas/grafeas/proto/v1beta1/common_go_proto"
+	pb "github.com/grafeas/grafeas/proto/v1beta1/grafeas_go_proto"
+	"github.com/pkg/errors"
+	"github.com/tektoncd/chains/pkg/artifacts"
+	"github.com/tektoncd/chains/pkg/chains/formats"
+	"github.com/tektoncd/chains/pkg/config"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/status"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+const (
+	StorageBackendGrafeas = "grafeas"
+	ProjectNameFormat     = "projects/%s"
+	NoteNameFormat        = "projects/%s/notes/%s"
+)
+
+// Backend is a storage backend that stores signed payloads in the storage that
+// is built on the top of grafeas i.e. container analysis.
+type Backend struct {
+	logger         *zap.SugaredLogger
+	tr             *v1beta1.TaskRun
+	client         pb.GrafeasV1Beta1Client
+	cfg            config.Config
+	occurrenceRefs []string // store occurrence IDs that are automatically generated during the time of creation
```

Comment:
I still think we need to change this behavior - we should be able to use the lookup functions without needing to statefully store the written occurrences at write time.

Reply:
Thanks for proposing this. Yes this indeed needs to be changed for the change of creating backends `per process` in which we'll not be able to store data statefully for a specific taskrun.

Here is the commit 14d2f37c8240d8f0bdadb5d813a46db727650f2e